h/security/policy/combined.py

```diff
@@ -56,38 +46,28 @@ def identity(self, request) -> Optional[Identity]:
         return self._identity_cache.get_or_create(request)
 
     def _load_identity(self, request):
-        return self._call_sub_policies("identity", request)
+        return call_policies("identity", request)
 
-    def _call_sub_policies(self, method, request, *args, **kwargs):
-        """
-        Delegate calls to the correct set of security policies.
 
-        :param method: Method to call (like `identity()` or `forget()`)
-        :param request: Pyramid request object
-        :param args: Args to pass to the method
-        :param kwargs: Kwargs to pass to the method
-        :return: The response from the correct sub-policy
-        """
+def call_policies(method: str, request, *args, **kwargs):
+    """
+    Call `method` on each applicable security policy and return the first result.
+
+    Call `method` on each security policy that is applicable to `request` in
+    turn. Return the result from the first policy that returns a truthy value
+    or the falsey result from the last policy that was called.
+    """
+    policies = [RemoteUserPolicy, CookiePolicy, BearerTokenPolicy, AuthClientPolicy]
```

Comment:
This is the new hard-coded order in which the four sub-policies will always be consulted.

In practice the only possible situations are:

1. It's an API request and `AuthClientPolicy.handles()` is truthy: `BearerTokenPolicy` will be called first then `AuthClientPolicy`. (`AuthClientPolicy.handles()` already existed before this PR and was already being consulted by the previous `_call_sub_policies()` method.)
2. It's an API request and `AuthClientPolicy.handles()` is falsey: only `BearerTokenPolicy` will be called
3. It's not an API request and `PROXY_AUTH` is truthy: only `RemoteUserPolicy` will be applicable
4. It's not an API request and `PROXY_AUTH` is falsey: only `CookiePolicy` will be applicable

These are the same four possible situations as previously.

If you look closely at [the previous `_call_sub_policies()` method](https://github.com/hypothesis/h/blob/ff26dc702f407a37a433e9c49a691d1fa30bc2d4/h/security/policy/combined.py#L61-L86) it either called `RemoteUserPolicy` only, `CookiePolicy` only, `BearerTokenPolicy` only or `BearerTokenPolicy` followed by `AuthClientPolicy`. So the explicitly hard-coded order here is the same as the previous implicit order (the only thing that matters is that `BearerTokenPolicy` comes before `AuthClientPolicy`).

Reply:
I spotted an unintentional change here: the previous `SecurityPolicy` [creates one instance of each sub-policy](https://github.com/hypothesis/h/blob/05a240e720ac433ff4c81f1145d8b495e735ebde/h/security/policy/combined.py#L31-L35) when it's initialized and re-uses them. The new `call_policies()` function creates new instances of the sub-policies each time it's called. Whether this is consequential I'm not sure but at least `BearerTokenPolicy` has [some state on `self`](https://github.com/hypothesis/h/blob/495c21a38d28787528b31247a26904176eff420b/h/security/policy/bearer_token.py#L23) so it might be. We might want to change it so that `SecurityPolicy` instantiates the sub-policies once (or probably better: the higher-level code that instantiates `SecurityPolicy` also instantiates the sub-policies and passes them into `SecurityPolicy.__init__()`) and then `SecurityPolicy` passes the list of them into `call_policies()`.